src/util/nextThought.js

```diff
@@ -0,0 +1,240 @@
+  const context = head(head(contextOf(contextChain)))
+  const contexts = getContextsSortedAndRanked(context.value)
+  const currentContextTop = head(contextChain)[0].value
+  const matchedContextChain = contexts.find(c => c.context.includes(currentContextTop))
+  const matchedContext = matchedContextChain.context.map(i => ({ value: i, rank: matchedContextChain.rank }))
+
+  return [...matchedContext, context, ...head(contextChain).slice(1)]
+}
+
+/**
+ * Computes the value of next thoughts  and context chain if-
+ * 1) the focused thought has context view open
+ * 2) the focus is on a context
+ * Delegates control to nextInThoughtView if none of the above conditions meet
+ * @param {string} value - The value of focused thought
+ * @param {string} rank - Rank of focused thought
+ * @param {array} path - Path to focsued thought
+ * @param {array} rankedContext - Context with rank
+ * @param {array} contextChain - ContextChain for the focused thought
+ * @param {(boolean | undefined)} ignoreChildren - Used to ignore the children context if they've been traversed already
+*/
+const nextInContextView = (value, rank, path, rankedContext, contextChain, ignoreChildren) => {
+  const { showHiddenThoughts, thoughtIndex } = store.getState()
+
+  if (rankedContext.length === 0 || path.length === 0) return null
+
+  const context = pathToContext(rankedContext)
+  const firstChild = perma(() => firstChildOfContextView(path, thoughtIndex))
+
+  // if the focus is on a thought with context view open, move it into context view - jump in
+  if (!ignoreChildren && isValidContextView(path) && firstChild()) {
+    const currentThought = head(path)
+    // jump out if there are no context children
+    return {
+      nextThoughts: [...firstChild(), currentThought],
+      contextChain
+    }
+  }
+  // if the focus is on or within a context
+  else if (isValidContextView(rankedContext)) {
+    const firstChild = perma(() => firstChildOfThoughtView(getPathFromContextChain(contextChain) || RANKED_ROOT, showHiddenThoughts))
+
+    const nextSibling = nextSiblingContext(rank, context, thoughtIndex, showHiddenThoughts)
+    const rankedContextHead = head(rankedContext)
+
+    return !ignoreChildren && firstChild() ? {
+      nextThoughts: unroot([...path, firstChild()]),
+      contextChain: []
+    }
+      : nextSibling ? {
+        nextThoughts: [...nextSibling, rankedContextHead],
+        contextChain: contextOf(contextChain)
+      }
+      : nextInThoughtView(rankedContextHead.value, contextOf(context), rankedContextHead.rank, contextOf(path), contextOf(contextChain), true)
+  }
+  // not a valid context view. Try navigating thoughts
+  else {
+    const contextWithNoChildren = contextChain.length === 1 &&
+      isContextViewActive(contextChain[0]) &&
+      getContexts(head(pathToContext(contextChain[0]))).length <= 1
+
+    return nextInThoughtView(value, context, rank, path, contextChain, contextWithNoChildren)
+  }
```

Comment:
In which cases does this `else` get executed? Is it possible to not call `nextInContextView` at all if this is the case, and instead call `nextInThoughtView` at the caller? I'm wondering if that might be cleaner.

If this is only hit when `nextThought` is called the first time, then we can definitely move some handling logic there and keep `nextInContextView` strictly for within a context view. If not, it might be more complicated, but I would like to understand the effort involved.